pennylane/hf/hamiltonian.py

```diff
@@ -352,6 +353,54 @@ def _generate_qubit_operator(op):
     return c, o
 
 
+def _simplify(h, cutoff=1.0e-12):
+    r"""Add together identical terms in the Hamiltonian.
+
+    The Hamiltonian terms with identical Pauli words are added together and eliminated if the
+    overall coefficient is zero.
+
+    Args:
+        h (Hamiltonian): PennyLane Hamiltonian
+        cutoff (float): cutoff value for discarding the negligible terms
+
+    Returns:
+        Hamiltonian: Simplified PennyLane Hamiltonian
+
+    **Example**
+
+    >>> c = np.array([0.5, 0.5])
+    >>> h = qml.Hamiltonian(c, [qml.PauliX(0) @ qml.PauliY(1), qml.PauliX(0) @ qml.PauliY(1)])
+    >>> print(_simplify(h))
+    (1.0) [X0 Y1]
+    """
+    wiremap = dict(zip(h.wires, range(len(h.wires) + 1)))
+
+    c = []
+    o = []
+    for i, op in enumerate(h.terms[1]):
+        op = qml.operation.Tensor(op).prune()
+        op = qml.grouping.pauli_word_to_string(op, wire_map=wiremap)
+        if op not in o:
+            c.append(h.terms[0][i])
+            o.append(op)
+        else:
+            c[o.index(op)] += h.terms[0][i]
+
+    coeffs = []
+    ops = []
+    nonzero_ind = np.argwhere(abs(np.array(c)) > cutoff).flatten()
+    for i in nonzero_ind:
+        coeffs.append(c[i])
+        ops.append(qml.grouping.string_to_pauli_word(o[i], wire_map=wiremap))
+
+    try:
+        coeffs = qml.math.stack(coeffs)
+    except ValueError:
+        pass
+
+    return qml.Hamiltonian(coeffs, ops)
+
+
```

Comment:
It would be great to reproduce this pattern in the `Hamiltonian` class. At the moment, `H.simplify()` is a method that manipulates the internal data of `H`, which is dangerous. We'd like it to return a new operator. Another option is `qml.simplify(H)`, which could be cool because other ops may be simplified in some way too.

I don't think this relates to this PR though, which is just about building Hamiltonians in a special way, right?

Reply:
This PR is literally adding a function that simplifies any input Hamiltonian. As far as I can tell, it should be possible to reproduce in the Hamiltonian class, or as you suggest, having `qml.simplify(H)`. I really like this latter suggestion actually

Reply:
Hi @mariaschuld, @ixfoduap and @josh146. Here are my thoughts on the simplify function.

This is a general functions that does not necessary belong to `qml.hf`. However, the way it is written currently is based on one important point: convert Pauli words to strings and manipulate the strings in with python. For example, instead of looping over the words and using `compare`, one can simply use 
`if o in ['XYZ', ...]`
However, I am not sure we want to manipulate Hamiltonians/operators by turning them to strings. Another philosophy could be generating the binary represantation of words and Hamiltonians and manipulate those objects numerically to, for example, compare or multiply them. I think we need to decide first what approach we want to follow for operator manipulation/arithmetic in general. This is the only reason that I hided `_simplify` inside `hf.hamiltonian`.

So, I want your thoughts on these options:
1. Convert `hf.hamiltonian._simplify` to `qml.simplify(H)`
2. Wait until we have an ADR on how we really want to _efficiently_ work with operators/Hamiltonian in PL and then create `qml.simplify(H)` accordingly.

Reply:
@soranjh I think these are very valid points, still a lot needs to be figured out in terms of how to best implement comparison/simplification long-term.

Having said that, it's not quite clear _when_ this research will be done/implemented. Since your version already is an improvement over what is already in PL, I am very much in favour of making it more general now, and continuing to improve it iteratively down the line.

In other words: since this is already a net benefit over what PL users have access to, I don't think there is much disadvantage to making this generally available in the short term?

Reply:
@soranjh I would suggest to go for 2 - this is the least work for you now, and it incurs only 100% non-user-facing changes in the future! You are right, Hamiltonian simplification needs an ADR. Also remember that all operators will optionally define terms, and the simplify function should be rather general!

Reply:
@mariaschuld makes the following proposal, which I think makes a lot of sense:

1. Continuing working on this PR

2. Once this PR is merged, improve the existing `Hamiltonian.simplify()` method with this code

3. Finally, we can use what we learnt to (a) inform the ADR, and (b) help us decide any final UI/design decisions.